### PR TITLE
WIP: API to diff layer tree to track damage

### PR DESCRIPTION
**Pull request by knopp**
## Description

Replaces https://github.com/flutter/engine/pull/21299

This PR adds api to track damage rectangles during rendering. Usage is relatively simple, it introduces the following structure in `compositor_context.h`
```cpp
struct FrameDamage {
  // IN: Previous layer tree
  const LayerTree* prev_layer_tree = nullptr;

  // IN: Additional damage (accumulated for double / triple buffering)
  SkIRect additional_damage = SkIRect::MakeEmpty();

  // OUT: Damage
  Damage damage_out;
};

struct Damage {
  // This is the damage between current and previous frame;
  // When using double or triple buffering, this damage must be acumulated
  // for each "inactive" framebuffer and then be provided as additional_damage.
  SkIRect net_damage;

  // Reflects actual change to target framebuffer; This is net_damage + damage
  // previously acumulated for the framebuffer. If embedder supports partial
  // update, this is the region that needs to be updated.
  SkIRect framebuffer_damage;
};
```

This is a completely new implementation that does tree diffing instead of determining the damage outside of tree. It does require some changes to framework to diff correctly, but those should be relatively minor.

## Related Issues

https://github.com/flutter/flutter/issues/33939

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [X] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation.
- [ ] All existing and new tests are passing.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

This shouldn't be breaking change, if `frame_damage` is not provided (default). 
However it does prevent rendering of picture, physical shape and texture layers that are outside of cull rect. 

Did any tests fail when you ran them? Please read [handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

### CODE REVIEW

File: flow/layers/container_layer.cc [hunk 1]
```diff
@@ -10,6 +10,91 @@ namespace flutter {
...
+    }
+    --new_children_bottom;
+    --old_children_bottom;
+  }
+
+  // old layers that don't match
+  for (int i = old_children_top; i <= old_children_bottom; ++i) {
+    auto layer = prev_layers[i];
+    DiffContext::AutoSubtreeRestore subtree(context);
+    context->MarkSubtreeDirty(layer->paint_region());
```
Comment (flar) [inline flow/layers/container_layer.cc:69]: Would this be easier if AddDamage were public? It seems to me that doing an autorestore per removed child is more work than we really need here.

File: flow/layers/container_layer.cc [hunk 2]
```diff
@@ -10,6 +10,91 @@ namespace flutter {
 
 ContainerLayer::ContainerLayer() {}
 
+void ContainerLayer::AssignOldLayer(std::shared_ptr<ContainerLayer> old_layer) {
+  old_layer->old_layer_ = nullptr;
```
Comment (flar) [inline flow/layers/container_layer.cc:14]: Perhaps once the Diff is done we can delete the old_layers so that they are more aggressively reclaimed? Maybe Diff could tail-nullify the pointers (i.e just before it returns from processing a subtree).

Comment (knopp) [reply flow/layers/container_layer.cc:14]: There is slight issue here. The old_layer_ is currently only used in [ContainerLayer::canDiff](https://github.com/flutter/engine/pull/21409/commits/a091548da47447d27028cd0b342fa48ec6d032d2#diff-125412910bbb6cc564b2a59b2cb3b1a9R20). The argument for actual diffing ([ContainerLayer::Diff](https://github.com/flutter/engine/pull/21409/commits/a091548da47447d27028cd0b342fa48ec6d032d2#diff-125412910bbb6cc564b2a59b2cb3b1a9R24))is const and it comes from layer above. 

I'm not even sure it's safe actually, if preroll returns kResubmitFrame or kSkipAndRetryFrame, the layer tree is put in front of pipeline and rasterization will repeat.

Comment (flar) [reply flow/layers/container_layer.cc:14]: It can be more complicated than that. If the Pipeline is full we drop layer_trees entirely and so they would never be processed in the first place.

Also, a layer_tree can be sent to us by the framework while we are processing the previous layer_tree on the Rasterizer, so trying to manage these pointers from both the compositor_context and the rasterizer thread can cause problems.

In upcoming work, we are trying to shift to not having a pipeline at all. We will always have just one layer_tree waiting to be rasterized (unless the app has gone static). This means that we might discard 5 layer_trees in a row and then rasterize the 6th. Meanwhile a lot of "old_layer" pointers will have been lost.

So, at the time that a compositor context is working, the old_layer could be processed and have a repaint_region computed for it. Or it could be lying in wait and the repaint_region has not yet been calculated. Or it is lying in wait but we will replace it and it will never be used.

It might be best to do the diff operation in the compositor_context and then hand off a layer that has all of its diffs and repaint_regions computed to the pipeline? If the pipeline replaces one waiting layer_tree with another then it should incorporate the damage from the replaced tree into the new one.

OR...

On the other hand, if we don't remove the old_layer->old_layer here and instead we remove it at the time that we do the Diff on the rasterizer thread then we won't have a race condition. If frames have been skipped then old_layer->old_layer would not be null and so we could do a "2[3,4,..N] deep" diff by following the "old_layer" chains to the end. A missing "old_layer" at the end of the chain means that that particular old layer has been diffed and we can trust its paint_region.

Comment (flar) [reply flow/layers/container_layer.cc:14]: I think my prototype was suffering from similar issues. I was doing some work in the compositor_context push*() routines to try to prepare for doing the diff later, but I would end up doing things like clearing the children on the old layer when I reused it (save the old children aside, make a new vector for the new children) which would have had drastic consequences if the previous layer hadn't been rasterized yet.

I'm beginning to lean towards the concept of keeping the old_layer pointers until we successfully diff a tree and then severing them at that time. Such a tree that is missing its old_layer pointers is a milestone with which we can diff a future tree against.

TBD - how that mechanism works when we have A/B/A/B alternating frame buffers.

Comment (knopp) [reply flow/layers/container_layer.cc:14]: > TBD - how that mechanism works when we have A/B/A/B alternating frame buffers

This is not really an issue. The solution here works with double buffering (A/B/A/B) and triple buffering (A/B/C/A/B/C). The way this works is that when you render to framebufer A, you take the net diff (D) and store it for framebuffer B. Then when you render to framebuffer B, you make sure, that you also include area D. 

It's late here, maybe the code will explain it better. It is from the prototype I'm testing metal with (triple buffering):
```cpp
    static std::map<int/*frame_buffer_id*/, SkIRect> damage_accumulator;

    // returns existing extram_damage or SkIRect::MakeSize(layer_tree.frame_size()) if none exsits
    SkIRect extra_damage =
        damage_accumulator
            .insert(std::make_pair(current_frame_buffer_id, 
                                   SkIRect::MakeSize(layer_tree.frame_size())))
            .first->second;
    FrameDamage frame_damage;
    frame_damage.prev_layer_tree = last_layer_tree_.get();

    // when rendering current frame, make sure that extra_damage area is covered as well 
    // this is the area where current framebuffer "lags behind" front buffer
    frame_damage.additional_damage = extra_damage;

    RasterStatus raster_status =
        compositor_frame->Raster(layer_tree, false, &frame_damage);

    for (auto& e : damage_accumulator) {
      if (e.first == current_frame_buffer_id) {
        // for current framebuffer reset accumulated damage
        e.second = SkIRect::MakeEmpty();
      } else {
        // for other framebuffers add accumulated damage
        e.second.join(frame_damage.damage_out.net_damage);
      }
    }
```
This will work with double, triple, n-buffering, as long as you have unique ID for each framebuffer. There really isn't need to diff with anything other than last layer tree. Also I don't think following old_layer->old_layer n times will give you tree n generations back, because some branches may be retained.

> I was doing some work in the compositor_context push*

I assume you mean scenebuilder. I indeed do clean old_layer->old_layer reference there. You may be right that this is not the best place for it. I couldn't see any raciness, but that's probably just because I'm only clearing one reference so there's lot less potential. This definitely needs to be addressed!

I still think though that ideally you want to rasterize every layertree that the framework produces. Dropping any single one would cause a jank, and when that happens, all bets are off and trying to do a diff seems pretty pointless.

Comment (flar) [reply flow/layers/container_layer.cc:14]: The race conditions would only happen if we have a long rasterization that takes more than a frame to pump the next layer_tree. If we always have `build, render, vsync, build, render, vsync` then we'd never see this. When that render cross the boundary to the next vsync interval then we still might not see it because it would have completed the diff/preroll before the vsync (which triggers the next build). It would only be a problem if we drop more than one frame, or if it gets around to rendering the saved layer_tree so late that the next one is being built during/while the saved one is being diff/prerolled.

Comment (knopp) [reply flow/layers/container_layer.cc]: I think the easiest way around the race condition is [this](https://github.com/flutter/engine/pull/21409/commits/187f856c0a7421329a32d64a9daea80941ec4a90).

The `old_layer_` is only used during [ContainerLayer::CanDiff(const Layer*)](https://github.com/flutter/engine/pull/21409/commits/187f856c0a7421329a32d64a9daea80941ec4a90#diff-125412910bbb6cc564b2a59b2cb3b1a9R21). So it should be perfectly fine to be stored as weak_ptr and be around for only as long as the previous layer tree exists.

Comment (flar) [reply flow/layers/container_layer.cc]: There is still the issue that if an old_layer is stored there, then it is only as valid as whether or not it was involved in the previous scene diff (the scene may have been replaced before the old scene was involved in a Diff) and whether or not that scene was rendered to a frame buffer (a scene could be Diff'd and Preroll'd and then aborted due to configuration changes detected during Preroll - the scene should be pushed back in that case, but if a new scene comes in while all of that is happening then it may end up getting discarded after it was Diff'd).

Comment (knopp) [reply flow/layers/container_layer.cc]: And that should be fine. In that case CanDiff will simply return false and we basically repaint the whole scene. At that point something bad enough has happened that we missed rasterization a frame, so repainting the whole layer tree doesn't seem like much of an issue.

Comment (flar) [reply flow/layers/container_layer.cc]: old_layer exists so why would CanDiff return false?

The issue won't manifest by old_layer getting reclaimed, it will manifest by old_layer existing, but not being the "last layer that was rendered".

Consider the following.

- An app constructs a scene with LayerA. The engine Diffs and starts rendering that scene.
- While that scene is being rendered another vsync happens so we build a new scene.
- The new scene has LayerB in place of LayerA. They differ by a property value so they represent a dirty region. EngineLayerB at the engine level points at EngineLayerA, but CanDiff would discover that it was dirty.
- While the previous scene with LayerA is still being rendered, another vsync happens and we build another new scene.
- The new scene reuses LayerB because that part of the scene didn't change. A new EngineLayerB2 points at its previous copy EngineLayerB.
- The rendering of the first scene is now done and the renderer starts rendering the latest scene.
- During Diff it finds EngineLayerB2 in the scene and it has a pointer to EngineLayerB and they have identical properties and so it believes that nothing has changed there.

The problem here is that EngineLayerA is the last version of that layer that was rendered, but it was dropped on the floor and there is no record of its historical presence in the latest layer_tree.

Comment (knopp) [reply flow/layers/container_layer.cc]: > The problem here is that EngineLayerA is the last version of that layer that was rendered, but it was dropped on the floor and there is no record of its historical presence in the latest layer_tree.

I don't follow this. We always diff current layer tree against `last_layer_tree_`. `last_layer_tree_` is the [last sucessfully rasterized](https://github.com/flutter/engine/blob/master/shell/common/rasterizer.cc#L341) layer tree. If EngineLayerA is the last version of that layer that was rendered, then it must be part of `last_layer_tree_`.

So during your diff, on one side, you'll have EngineLayerA (from `last_layer_tree_`), and on other side you'll have EngineLayerB2 (with old_layer_ = EngineLayerB).

So the comparison is EngineLayerB2->CanDiff(EngineLayerA), which turns into EngineLayerB2 == EngineLayerA || EngineLayerB2->old_layer_ == EngineLayerA. both of which are false. So we'll re-render the subtree.

Comment (knopp) [reply flow/layers/container_layer.cc]: All this said, if we have a situation like this:

last rasterized  LayerTreeA (=`last_layer_tree_`),
discarded LayerTreeB1 ... LayerTreeBn 
and currently being rasterized LayerTreeC

the diff between LayerTreeC and LayerTreeA in this implementation would be pretty much full repaint, even though they might have things in common. If we however write `ContainerLayer::CanDiff` like this:

```cpp
  bool ContainerLayer::CanDiff(const Layer* layer) const {
    const ContainerLayer* l = this;
    while (l) {
      if (l == layer) {
        return true;
      }
      l = reinterpret_cast<const ContainerLayer*>(l->old_layer_.lock().get());
    }
    return false;
  }
```
And then keep reference to LayerTreeB1...LayerTreeBn until we successfully rasterize LayerTreeC, we could get an actual diff between LayerTreeC and LayerTreeA.

Comment (flar) [reply flow/layers/container_layer.cc]: Here is part of the confusion. There are 2 entirely separate things being called "old_layer" here.

ContainerLayer::old_layer_ which is a field in the container assigned from the `push...()` methods in scene_builder.
ContainerLayer::Diff(context, old_layer) which is passed down from the previous rasterization.

I missed that there were two different layer pointers being referred to as "old_layer" when I was reading the code, so that drove most of my confusion. Better naming would help a lot here.

So, CanDiff is basically asking the question of whether we are rasterizing the same identical layer or whether we are rasterizing the layer that directly replaced the (same identical) layer. I can see that logic now.  Yes, if the layer is lost to the mists of time then we would just miss out on an opportunity for optimization.

We could also just store the ID so we don't have to keep a pointer. layer1 == layer2 should be identical to layer1.ID == layer2.ID. If we store the id then we no longer need to worry about keeping old layer storage alive.

With respect to chaining, assign could do something like this for a guaranteed history of 2 distinct layers for free:

```
AssignOldLayer(Layer *replaced_layer) {
  if (replaced_layer->uniqueID() == this->uniqueID()) {
    // We weren't replaced in this current scene, so transfer forward the ID of the most recently replaced layer.
    this.replaced_ID = replaced_layer.replaced_ID;
  } else {
    // replaced_layer becomes the newest most recently replaced layer so copy forward its ID.
    this.replaced_ID = replaced_layer->uniqueID();
  }
}
```

Basically replaced_ID will be the ID of the last layer which was in this same spot, but which had a different ID. Would that open up more optimization opportunities?

Comment (flar) [reply flow/layers/container_layer.cc]: In my pseudo-code in my previous comment, the test could also just be `if (replaced_layer == this)` I think. That comparison doesn't need to be done by ID.

Comment (knopp) [reply flow/layers/container_layer.cc]: I like the ID idea a lot! I [tried it](https://github.com/flutter/engine/pull/21409/commits/44ba8a38a11bd9ccf8614deffdb1bf325fd0fcfd) and it seems to be working well.

Each layer starts with `diff_id_ = unique_id_`, and then in `AssignOldLayer` the `diff_id_` is simply replaced by `old_layer->diff_id_`.

This way the `diff_id_` is basically propagated `unique_id_` of the first layer in "chain" of replacement layers that can be diffed. No storage required and `last_layer_tree_` can be properly diffed with current tree regardless of whether there have been any discarded intermediate trees.

Comment (flar) [reply flow/layers/container_layer.cc]: GMTA... I was just logging in to mention that trick. This way we can have an infinity history of equivalence. Woot!

Comment (flar) [reply flow/layers/container_layer.cc]: diff_id? or equivalence_id? or origin_id? Not sure what word best describes this...

Comment (knopp) [reply flow/layers/container_layer.cc]: Yeah, diff_id is a rather generic name. I quite like origin_id. It seems to express the fact that there was first layer somewhere in chain that had this id. Perhaps also original_layer_id.

Comment (flar) [reply flow/layers/container_layer.cc]: Those sound good. I also thought that this ID creates an equivalence relation. But origin is more descriptive of where it came from.

`first_apperance_of_this_layer_in_the_scene_ID` is probably too long...

**Comment by knopp** [comment]
Feedback has been worked in. Replaced with https://github.com/flutter/engine/pull/21498.
